prow/cmd/hmac/main.go

```diff
@@ -0,0 +1,389 @@
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
+	}
+
+	kc, err := o.kubernetes.InfrastructureClusterClient(o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating Kubernetes client for infrastructure cluster.")
+	}
+
+	currentHmacYaml, err := getCurrentHmacTokens(kc, o.hmacTokenSecretNamespace, o.hmacTokenSecretName, o.hmacTokenKey)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting the current hmac yaml.")
+	}
+
+	var currentHMACMap map[string]github.HmacsForRepo
+	if err := yaml.Unmarshal(currentHmacYaml, &currentHMACMap); err != nil {
+		logrus.WithError(err).Error("Couldn't unmarshal the hmac secret as hierarchical file. Parsing as a single global token and writing it back to the secret.")
+		currentHMACMap["*"] = github.HmacsForRepo{
+			github.HmacSecret{
+				Value: string(currentHmacYaml),
+			},
+		}
+	}
+
+	var configAgent config.Agent
+	if err := configAgent.Start(o.configPath, ""); err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+	newHMACConfig := configAgent.Config().ManagedWebhooks
+
+	c := client{
+		kubernetesClient: kc,
+		options:          o,
+
+		currentHMACMap: currentHMACMap,
+		newHMACConfig:  newHMACConfig,
+	}
+
+	if err := c.handleConfigUpdate(); err != nil {
+		logrus.WithError(err).Fatal("Error handling hmac config update.")
+	}
+}
+
+func (c *client) handleConfigUpdate() error {
+	repoAdded := map[string]config.ManagedWebhookInfo{}
+	repoRemoved := map[string]bool{}
+	repoRotated := map[string]config.ManagedWebhookInfo{}
+
+	for repoName, hmacConfig := range c.newHMACConfig {
+		if _, ok := c.currentHMACMap[repoName]; ok {
+			repoRotated[repoName] = hmacConfig
+		} else {
+			repoAdded[repoName] = hmacConfig
+		}
+	}
+
+	for repoName := range c.currentHMACMap {
+		if _, ok := c.newHMACConfig[repoName]; !ok {
+			repoRemoved[repoName] = true
+		}
+	}
+
+	// Remove the webhooks for the given repos, as well as removing the tokens from the current hmac map.
+	if err := c.handleRemovedRepo(repoRemoved); err != nil {
+		return fmt.Errorf("error handling hmac update for removed repos: %v", err)
+	}
+
+	// Generate new hmac token for required repos, do batch update for the hmac token secret,
+	// and then iteratively update the webhook for each repo.
+	if err := c.handleAddedRepo(repoAdded); err != nil {
+		return fmt.Errorf("error handling hmac update for new repos: %v", err)
+	}
+	if err := c.handledRotatedRepo(repoRotated); err != nil {
+		return fmt.Errorf("error handling hmac rotations for the repos: %v", err)
+	}
+	// Update the hmac token secret first, to guarantee the new tokens are available to hook.
+	if err := c.updateHmacTokenSecret(); err != nil {
+		return fmt.Errorf("error updating hmac tokens: %v", err)
+	}
+	// HACK: sleep 20 seconds to wait for the reconciliation to finish.
```

Comment:
nit: In particular we are waiting for the `hmac` k8s secret update to propagate to the pods that are using the secret so that components like `hook` can start respecting the new hmac values.

Reply:
Updated the comment as suggested. Thanks!